### PR TITLE
ENH: refactor architecture of AlignedSeqsData and changes to ABC

### PR DESCRIPTION
**Pull request by GavinHuttley**
## Summary by Sourcery

Refactor the architecture of AlignedSeqsData to enhance sequence data management and improve efficiency. Update the method for retrieving sequence lengths and revise tests to align with the new implementation. Introduce new tests for sequence initialization with annotation databases.

Enhancements:
- Refactor the architecture of AlignedSeqsData to improve data handling and sequence management.
- Modify the AlignedSeqsData class to use a more efficient internal representation for sequence data, including changes to how gaps and ungapped sequences are managed.
- Update the method for obtaining sequence lengths to a more efficient implementation using get_seq_length.

Tests:
- Revise tests to accommodate changes in the AlignedSeqsData class, including updates to test cases for sequence length retrieval and sequence array handling.
- Add new tests for initializing sequences with annotation databases and offsets.

### CODE REVIEW

File: src/cogent3/core/new_alignment.py [hunk 1]
```diff
@@ -3011,38 +3042,38 @@ class AlignedSeqsData(AlignedSeqsDataABC):
     def __init__(
         self,
         *,
-        seqs: Optional[dict[str, StrORBytesORArray]],
-        gaps: Optional[dict[str, numpy.ndarray]],
+        gapped_seqs: numpy.ndarray,
```
Comment (KatherineCaley) [inline src/cogent3/core/new_alignment.py:3045]: refactor comment that we need a docstring describing construction

File: src/cogent3/core/new_alignment.py [hunk 2]
```diff
@@ -3011,38 +3042,38 @@ class AlignedSeqsData(AlignedSeqsDataABC):
...
+        self._ungapped = ungapped_seqs or {}
+        self._gaps = gaps or {}
+        align_len = align_len or gapped_seqs.shape[1]
+        if align_len:
+            assert align_len == gapped_seqs.shape[1], "mismatch in alignment length"
+
+        self._align_len = align_len
         self._offset = offset or {}
 
+        if check:
```
Comment (KatherineCaley) [inline src/cogent3/core/new_alignment.py:3067]: we should add a check that the number of names matches the shape of the gapped array

**Review by KatherineCaley — APPROVED** [review]
looks good!

File: src/cogent3/core/new_alignment.py [hunk 3]
```diff
@@ -3011,38 +3042,72 @@ class AlignedSeqsData(AlignedSeqsDataABC):
...
+        offset
+            a dictionary of annotation offsets
+        align_len
+            length of the alignment, which must equal the gapped_seqs.shape[1]
+        check
+            validate any keys in offset, ungapped_seqs, gaps are a subset of names
+
+        Raises
+        ------
+        ValueError
```
Comment (KatherineCaley) [inline src/cogent3/core/new_alignment.py:3080]: boiler plate code?

File: src/cogent3/core/new_alignment.py [hunk 4]
```diff
@@ -3011,38 +3042,72 @@ class AlignedSeqsData(AlignedSeqsDataABC):
...
+            2D numpy.uint8 array of aligned sequences. axis 0 are sequences,
+            axis 1 are alignment positions
+        names
+            sequence names in order matching the axis 0 of gapped_seqs
+        alphabet
+            caharacter alphabet for the sequences
+        ungapped_seqs
+            a dictionary mapping names to 1D numpy.uint8 arrays of individual
+            sequences without gaps. If not provided, computed on demand.
+        gaps, optional
```
Comment (KatherineCaley) [inline src/cogent3/core/new_alignment.py:3067]: the comment that gaps is optional is not consistent with the rest of the docstring

File: src/cogent3/core/new_alignment.py (hunk 3 above)
Comment (GavinHuttley) [reply src/cogent3/core/new_alignment.py:3080]: yup, missed that!

**Pull request merged**